File: iomerge-server-app/src/main/java/com/github/krzychek/server/network/EventServer.java
package com.github.krzychek.server.network;

import com.github.krzychek.server.api.appState.AppState;
import com.github.krzychek.server.api.appState.AppStateManager;
import org.annoprops.annotations.ConfigProperty;
import org.annoprops.annotations.PropertyHolder;
import org.pmw.tinylog.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;


/**
 * Waits for clients and starts up connection
 */
@PropertyHolder
@Component
public class EventServer {

	private final AppStateManager appStateManager;
	private final ConnectionHandlerProxy connectionHandlerProxy;
	private final Executor executor;

	private ServerSocket serverSocket;

	@ConfigProperty("ServerPort")
	private int port = 7698;

	@Autowired
	public EventServer(AppStateManager appStateManager, ConnectionHandlerProxy connectionHandlerProxy) {
		this.appStateManager = appStateManager;
		this.connectionHandlerProxy = connectionHandlerProxy;
		this.executor = Executors.newSingleThreadExecutor();
	}

	@PreDestroy
	private void shutdown() {
		Logger.info("shutting down server");

		if (serverSocket != null) try {
			serverSocket.close();
		} catch (IOException e) {
			Logger.warn(e);
		}

		appStateManager.disconnected();
	}

	@PostConstruct
	public void start() throws IOException {
		Logger.info("starting server at port " + port);
		serverSocket = new ServerSocket();
		serverSocket.setPerformancePreferences(1, 2, 0);
		serverSocket.bind(new InetSocketAddress(port));
		Logger.info("listening at port " + port);
	}

	private void restart() {
		Logger.info("restarting server");
		shutdown();
		try {
			start();
		} catch (IOException e) {
			Logger.error(e);
			throw new RuntimeException(e);
		}
	}

	private void acceptListener() {
		try {
			Socket clientSocket = serverSocket.accept();
			Logger.info("client socket accepted");
			connectionHandlerProxy.connect(clientSocket);

		} catch (IOException e) {
			Logger.warn("Problem with connection", e);
		}
	}

	@EventListener
	private void onAppStateChange(AppState appState) {
		if (AppState.DISCONNECTED.equals(appState)) {
			executor.execute(this::acceptListener);
		}
	}

	public int getPort() {
		return port;
	}

	public void setPort(int port) {
		if (this.port == port) return;
		this.port = port;
		if (serverSocket != null && serverSocket.isBound()) restart();
	}
}

File: iomerge-server-app/src/main/java/com/github/krzychek/iomerge/server/network/EventServer.kt
package com.github.krzychek.iomerge.server.network

import com.github.krzychek.iomerge.server.api.appState.AppState
import com.github.krzychek.iomerge.server.api.appState.AppStateManager
import com.google.common.eventbus.Subscribe
import org.annoprops.annotations.ConfigProperty
import org.annoprops.annotations.PropertyHolder
import org.pmw.tinylog.Logger
import org.springframework.stereotype.Component
import java.io.EOFException
import java.io.IOException
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.util.concurrent.Executors
import javax.annotation.PreDestroy


/**
 * Waits for clients and starts up connection
 */
@PropertyHolder
@Component
open class EventServer(private val appStateManager: AppStateManager, private val connectionHandlerProxy: ConnectionHandlerProxy) {
	private val executor = Executors.newSingleThreadExecutor()

	private var serverSocket: ServerSocket? = null

	@ConfigProperty("ServerPort")
	var port = 7698
		set(port) {
			if (field != port) {
				field = port
				if (serverSocket != null) restart()
			}

		}

	@PreDestroy
	private fun shutdown() {
		Logger.info("shutting down server")

		try {
			serverSocket?.close()
			serverSocket = null
		} catch (e: IOException) {
			Logger.warn(e)
		}

		appStateManager.disconnected()
	}

	private fun restart() {
		Logger.info("restarting server")
		shutdown()
		acceptListener()
	}

	private fun acceptListener() = executor.execute {
		val serverSocket = ServerSocket().apply {
			setPerformancePreferences(1, 2, 0)
			bind(InetSocketAddress(port))
		}
		this.serverSocket = serverSocket
		Logger.info("listening at port " + this.port)

		try {
			val clientSocket = serverSocket.accept()
			Logger.info("client socket accepted")
			connectionHandlerProxy.connect(clientSocket)

		} catch (e: EOFException) {
			Logger.info("Problem with connection", e)
		} catch (e: IOException) {
			Logger.warn("Problem with connection", e)
		}
	}


	@Subscribe
	fun onAppStateChange(appState: AppState) {
		if (AppState.DISCONNECTED == appState) {
			this.acceptListener()
		}
	}
}
